InvenTree/users/models.py

```diff
@@ -1 +1,314 @@
+        (permission, app, model) = p.natural_key()
+
+        permission_string = '{app}.{perm}'.format(
+            app=app,
+            perm=permission
+        )
+
+        group_permissions.add(permission_string)
+
+    # List of permissions which must be added to the group
+    permissions_to_add = set()
+
+    # List of permissions which must be removed from the group
+    permissions_to_delete = set()
+
+    def add_model(name, action, allowed):
+        """
+        Add a new model to the pile:
+
+        args:
+            name - The name of the model e.g. part_part
+            action - The permission action e.g. view
+            allowed - Whether or not the action is allowed
+        """
+
+        if action not in ['view', 'add', 'change', 'delete']:
+            raise ValueError("Action {a} is invalid".format(a=action))
+
+        permission_string = RuleSet.get_model_permission_string(model, action)
+
+        if allowed:
+
+            # An 'allowed' action is always preferenced over a 'forbidden' action
+            if permission_string in permissions_to_delete:
+                permissions_to_delete.remove(permission_string)
+
+            if permission_string not in group_permissions:
+                permissions_to_add.add(permission_string)
+
+        else:
+
+            # A forbidden action will be ignored if we have already allowed it
+            if permission_string not in permissions_to_add:
+
+                if permission_string in group_permissions:
+                    permissions_to_delete.add(permission_string)
+
+    # Get all the rulesets associated with this group
+    for r in RuleSet.RULESET_CHOICES:
+
+        rulename = r[0]
+
+        try:
+            ruleset = RuleSet.objects.get(group=group, name=rulename)
+        except RuleSet.DoesNotExist:
+            # Create the ruleset with default values (if it does not exist)
+            ruleset = RuleSet.objects.create(group=group, name=rulename)
+
+        # Which database tables does this RuleSet touch?
+        models = ruleset.get_models()
+
+        for model in models:
+            # Keep track of the available permissions for each model
+
+            add_model(model, 'view', ruleset.can_view)
+            add_model(model, 'add', ruleset.can_add)
+            add_model(model, 'change', ruleset.can_change)
+            add_model(model, 'delete', ruleset.can_delete)
+
+    def get_permission_object(permission_string):
+        """
+        Find the permission object in the database,
+        from the simplified permission string
+
+        Args:
+            permission_string - a simplified permission_string e.g. 'part.view_partcategory'
+
+        Returns the permission object in the database associated with the permission string
+        """
+
+        (app, perm) = permission_string.split('.')
+
+        (permission_name, model) = perm.split('_')
+
+        content_type = ContentType.objects.get(app_label=app, model=model)
```

Comment:
To avoid Django to error for any mislabeled app and model (human mistakes can happen), I suggest replacing with:
```
        try:
            content_type = ContentType.objects.get(app_label=app, model=model)
            permission = Permission.objects.get(content_type=content_type, codename=perm)
        except ContentType.DoesNotExist:
            permission = None
```

`group.permissions.add(permission)` does not error out if None object is passed.

Better yet would be to report this to the user in a "nice" way.